pkl-core/src/main/java/org/pkl/core/util/RrbTree.java

```diff
@@ -0,0 +1,2257 @@
+/*
+ * Copyright © 2016-2025 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.core.util;
+
+import static org.organicdesign.fp.collections.Cowry.*;
+import static org.organicdesign.fp.indent.IndentUtils.arrayString;
+import static org.organicdesign.fp.indent.IndentUtils.indentSpace;
+
+import java.io.*;
+import java.util.Arrays;
+import java.util.List;
+import org.organicdesign.fp.collections.BaseList;
+import org.organicdesign.fp.collections.ImList;
+import org.organicdesign.fp.collections.MutList;
+import org.organicdesign.fp.collections.PersistentVector;
+import org.organicdesign.fp.collections.UnmodIterable;
+import org.organicdesign.fp.collections.UnmodSortedIterable;
+import org.organicdesign.fp.collections.UnmodSortedIterator;
+import org.organicdesign.fp.function.Fn0;
+import org.organicdesign.fp.indent.Indented;
+import org.organicdesign.fp.oneOf.Option;
+import org.organicdesign.fp.tuple.Tuple2;
+import org.organicdesign.fp.tuple.Tuple4;
+
+// This file is adapted from https://github.com/GlenKPeterson/Paguro:
+//
+// Copyright 2016-05-28 PlanBase Inc. & Glen Peterson
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```

Comment:
Let's do what we've been doing so far: create a new package, and put this as a comment on that package